speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.cpp

```diff
@@ -98,64 +121,307 @@ void FileDataStorage::Shutdown()
+void FileDataStorage::WriteInterventionData()
+{
+    bool decisionMade = false;
+    WriteDecision(GetMedian(m_steerDecision), "SteerDecision", decisionMade);
+    WriteDecision(GetMedian(m_brakeDecision), "BrakeDecision", decisionMade);
+    WriteDecision(GetMedian(m_accelDecision), "AccelDecision", decisionMade);
+    WriteDecision(GetLeastCommon(m_gearDecision), "GearDecision", decisionMade);
+    WriteDecision(GetLeastCommon(m_lightDecision), "LightsDecision", decisionMade);
+
+    if (!decisionMade)
     {
-        WRITE_STRING(m_outputStream, "AccelDecision");
-        WRITE_VAR(m_outputStream, p_decisions.GetAccel());
+        WRITE_STRING(m_outputStream, "Decisions");
     }
-    if (p_decisions.ContainsGear())
+    WRITE_STRING(m_outputStream, "NONE");
+}
+
+/// @brief Writes the p_decisionType data from the last m_compressionRate time steps to the buffer file
+/// @param p_decision the value of the p_decisionType of the last time m_comrpessionRate time steps
+/// @param p_decisionType the current decision type
+/// @param p_decisionMade if there has already been a decision made
+template <typename TNumber>
+void FileDataStorage::WriteDecision(TNumber p_decision, const std::string& p_decisionType, bool& p_decisionMade)
+{
+    if (p_decision != -1)
     {
-        WRITE_STRING(m_outputStream, "GearDecision");
-        WRITE_VAR(m_outputStream, p_decisions.GetGear());
+        if (!p_decisionMade)
+        {
+            p_decisionMade = true;
+            WRITE_STRING(m_outputStream, "Decisions");
+        }
+        WRITE_STRING(m_outputStream, p_decisionType);
+        WRITE_VAR(m_outputStream, p_decision);
     }
-    if (p_decisions.ContainsLights())
+}
+
+/// @brief Add the new value to the total of the current time steps
+/// @param p_total The current total of the variable
+/// @param p_value The new value of this timestep for the variable
+void FileDataStorage::AddForAveraging(float& p_total, float p_value)
+{
+    p_total += p_value;
+}
+
+/// @brief Calculates the average of the past time steps for a variable and resets it
+/// @param p_total The current total of the variable
+/// @return The average of the past time steps for a variable
+float FileDataStorage::GetAverage(float& p_total) const
+{
+    float p_average = p_total / static_cast<float>(m_compressionRate);
+    p_total = 0;
+    return p_average;
+}
+
+/// @brief Add the new value to the array in the correct compression step
+/// @param p_values Array with values from the current compression step
+/// @param p_value The new value of this time step for the variable
+/// @param p_compressionStep The current compression step
+template <typename TNumber>
+void FileDataStorage::AddToArray(TNumber p_values[], TNumber p_value, unsigned long p_compressionStep) const
+{
+    int p_placeInArray = static_cast<int>(p_compressionStep % static_cast<unsigned long>(m_compressionRate));
+    p_values[p_placeInArray] = p_value;
+}
```

Comment:
Not a parameter, shouldn't be called p_placeInArray
```suggestion
void FileDataStorage::AddToArray(TNumber p_values[], TNumber p_value, unsigned long p_compressionStep) const
{
    int placeInArray = static_cast<int>(p_compressionStep % static_cast<unsigned long>(m_compressionRate));
    p_values[placeInArray] = p_value;
}
```